File: checks/python/src/section508/checks/accessible_name.py
from section508.dom import (
    SPACE,
    attr,
    attr_or,
    clone,
    closest,
    find_id,
    has_attr,
    label_for,
    parse_number,
    remove_matches,
    select,
    tag_name,
)

WIDGET_ROLES = {
    "button",
    "link",
    "checkbox",
    "radio",
    "tab",
    "menuitem",
    "switch",
    "combobox",
    "slider",
    "searchbox",
    "textbox",
}


def check(root):
    if not root:
        return []

    controls = find_controls(root)
    if not controls:
        return ["This section has no control to name."]

    for control in controls:
        if not has_name(control, root):
            return [f"“{hint(control)}” has no programmatic name."]
    return []


def find_controls(root):
    natives = []
    for element in select(root, "button, a[href], input, select, textarea, summary"):
        if tag_name(element) != "input":
            natives.append(element)
            continue
        if (attr(element, "type") or "text").lower() != "hidden":
            natives.append(element)

    native_ids = {id(element) for element in natives}
    fakes = []
    for element in select(root, "div, span, a, img"):
        if id(element) in native_ids:
            continue
        if closest(element, "button, a[href], label, summary"):
            continue
        if select(element, "button, a[href], input, select, textarea"):
            continue
        if has_attr(element, "onclick"):
            fakes.append(element)
            continue
        role = (attr(element, "role") or "").lower()
        if role in WIDGET_ROLES:
            fakes.append(element)
            continue
        tabindex = attr(element, "tabindex")
        number = parse_number(tabindex) if tabindex is not None else None
        if tabindex is not None and number is not None and number >= 0:
            fakes.append(element)
    return natives + fakes


def has_name(control, root) -> bool:
    aria = SPACE.sub(" ", attr(control, "aria-label") or "").strip()
    if aria:
        return True

    ids = [ref for ref in (attr(control, "aria-labelledby") or "").split() if ref]
    if ids:
        text = " ".join(find_id(root, ref).get_text() if find_id(root, ref) is not None else "" for ref in ids)
        if SPACE.sub(" ", text).strip():
            return True

    if tag_name(control) == "img":
        return (attr(control, "alt") or "").strip() != ""

    if tag_name(control) == "input":
        type_name = (attr(control, "type") or "text").lower()
        if type_name in {"submit", "button", "reset"}:
            if (attr(control, "value") or "").strip():
                return True
            return type_name in {"submit", "reset"}
        if type_name == "image":
            return (attr(control, "alt") or "").strip() != ""
        return has_label(control, root) or (attr(control, "title") or "").strip() != ""

    if tag_name(control) in {"select", "textarea"}:
        return has_label(control, root) or (attr(control, "title") or "").strip() != ""

    role = (attr(control, "role") or "").lower()
    native_named = tag_name(control) in {"button", "a", "summary"}
    if not native_named and role not in WIDGET_ROLES:
        return False

    if own_text(control):
        return True
    images = select(control, "img")
    return bool(images and (attr(images[0], "alt") or "").strip())


def has_label(control, root) -> bool:
    wrap = closest(control, "label")
    if wrap and label_text(wrap, control):
        return True
    control_id = attr(control, "id")
    if not control_id:
        return False
    label = label_for(root, control_id)
    return bool(label and SPACE.sub(" ", label.get_text()).strip())


def label_text(wrap, control):
    copied = clone(wrap)
    remove_matches(copied, "input, select, textarea, button")
    return SPACE.sub(" ", copied.get_text()).strip()


def own_text(element) -> str:
    copied = clone(element)
    remove_matches(copied, "script, style")
    return SPACE.sub(" ", copied.get_text()).strip()


def hint(control) -> str:
    return SPACE.sub(
        " ",
        (
            own_text(control)
            or attr_or(control, "aria-label")
            or attr_or(control, "id")
            or attr_or(control, "name")
            or tag_name(control)
        ),
    ).strip()

File: checks/ruby/lib/section508/checks/accessible_name.rb
# frozen_string_literal: true

module Section508
  module Checks
    # Every control has a programmatic name.
    class AccessibleName
      WIDGET_ROLES = %w[
        button link checkbox radio tab menuitem switch combobox slider searchbox textbox
      ].freeze

      def self.call(root)
        return [] unless root&.respond_to?(:css)

        controls = find_controls(root)
        return ["This section has no control to name."] if controls.empty?

        controls.each do |control|
          unless named?(control, root)
            return ["“#{hint(control)}” has no programmatic name."]
          end
        end
        []
      end

      def self.find_controls(root)
        natives = Dom.elements(root, "button, a[href], input, select, textarea, summary").select do |element|
          next true unless element.name.upcase == "INPUT"

          (element["type"] || "text").downcase != "hidden"
        end

        fakes = Dom.elements(root, "div, span, a, img").select do |element|
          next false if natives.include?(element)
          next false if Dom.closest(element, "button, a[href], label, summary")
          next false if element.at_css("button, a[href], input, select, textarea")
          next true if Dom.has_attr?(element, "onclick")

          role = element["role"].to_s.downcase
          next true if WIDGET_ROLES.include?(role)

          tabindex = element["tabindex"]
          !tabindex.nil? && number(tabindex) >= 0
        end

        natives + fakes
      end

      def self.named?(control, root)
        aria = control["aria-label"].to_s.gsub(/\s+/, " ").strip
        return true unless aria.empty?

        ids = control["aria-labelledby"].to_s.split(/\s+/).reject(&:empty?)
        if ids.any?
          text = ids.map { |id| node = Dom.by_id(root, id); node ? Dom.text(node) : "" }.join(" ")
          return true unless text.gsub(/\s+/, " ").strip.empty?
        end

        return !control["alt"].to_s.strip.empty? if control.name.upcase == "IMG"

        if control.name.upcase == "INPUT"
          type = (control["type"] || "text").downcase
          if %w[submit button reset].include?(type)
            return true unless control["value"].to_s.strip.empty?
            return type == "submit" || type == "reset"
          end
          return !control["alt"].to_s.strip.empty? if type == "image"

          return labeled?(control, root) || !control["title"].to_s.strip.empty?
        end

        if %w[SELECT TEXTAREA].include?(control.name.upcase)
          return labeled?(control, root) || !control["title"].to_s.strip.empty?
        end

        role = control["role"].to_s.downcase
        native_named = %w[BUTTON A SUMMARY].include?(control.name.upcase)
        return false unless native_named || WIDGET_ROLES.include?(role)

        return true unless own_text(control).empty?

        image = control.at_css("img")
        image && !image["alt"].to_s.strip.empty?
      end

      def self.labeled?(control, root)
        wrap = Dom.closest(control, "label")
        return true if wrap && !Dom.label_text(wrap, control).empty?

        id = control["id"]
        return false if id.nil? || id.empty?

        label = Dom.label_for(root, id)
        label && !Dom.norm(label).empty?
      end

      def self.own_text(element)
        clone = element.dup
        clone.css("script, style").each(&:remove)
        clone.text.to_s.gsub(/\s+/, " ").strip
      end

      def self.hint(control)
        (
          own_text(control).then { |text| text.empty? ? nil : text } ||
          control["aria-label"] ||
          control["id"] ||
          control["name"] ||
          control.name.downcase
        ).to_s.gsub(/\s+/, " ").strip
      end

      def self.number(value)
        Float(value)
      rescue ArgumentError, TypeError
        Float::NAN
      end
    end
  end
end
